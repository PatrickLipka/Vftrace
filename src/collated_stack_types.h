#ifndef COLLATED_STACK_TYPES_H
#define COLLATED_STACK_TYPES_H

#include <stdint.h>

#include <stdbool.h>

#include "stack_types.h"

typedef struct {
   // local stack
   stack_t *local_stack;
   // gloabl stackID
   int gid;
   bool precise;
   int caller;
   // need a copy of the name for even for local functions
   // because functions from other processes might
   // not appear in the local symbol table (e.g. two binaries)
   char *name;
   // hash
   uint64_t hash;
} collated_stack_t;

typedef struct {
   int nstacks;
   collated_stack_t *stacks;
} collated_stacktree_t;

#endif
