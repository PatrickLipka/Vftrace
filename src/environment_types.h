#ifndef ENVIRONMENT_TYPES_H
#define ENVIRONMENT_TYPES_H

#include <stdbool.h>

#include <regex.h>

typedef enum {
   env_none,
   env_int,
   env_long,
   env_longlong,
   env_float,
   env_double,
   env_bool,
   env_string,
   env_regex
} env_var_kind;

typedef union {
   int int_val;
   long long_val;
   long long longlong_val;
   float float_val;
   double double_val;
   bool bool_val;
   char *string_val;
   regex_t *regex_val;
} env_var_value;

typedef struct {
   env_var_kind value_kind;
   env_var_value value;
   char *value_string;
   char *name;
   bool set;
} env_var_t;

typedef struct {
   // in order to access the environment variables
   // like an array by index, all non-environment variables
   // need to be at the end of the struct!
   env_var_t vftrace_off;
   env_var_t do_sampling;
   env_var_t output_directory;
   env_var_t logfile_basename;
   env_var_t logfile_for_ranks;
   env_var_t ranks_in_mpi_profile;
   env_var_t sampletime;
   env_var_t mpi_log;
   env_var_t mpi_show_sync_time;
   env_var_t vfd_bufsize;
   env_var_t include_only_regex;
   env_var_t scenario_file;
   env_var_t preciseregex;
   env_var_t print_stack_profile;
   env_var_t print_stacks_for;
   env_var_t strip_module_names;
   env_var_t sort_profile_table;
   env_var_t sort_mpi_table;
   env_var_t show_overhead;
   env_var_t show_calltime_imbalances;
   env_var_t print_environment;
   env_var_t callpath_in_profile;
   env_var_t callpath_in_mpi_profile;
   env_var_t demangle_cxx;
   bool valid;
   int nenv_vars;
} environment_t;

#endif
