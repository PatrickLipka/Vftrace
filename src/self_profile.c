#include <stdio.h>

#ifdef _SELF_PROFILE_VFTRACE
char *vftr_self_prof_filename = NULL;
FILE *vftr_self_prof_iohandle = NULL;
char *vftr_self_prof_iobuffer = NULL;
#endif
