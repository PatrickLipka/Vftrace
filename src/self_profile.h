#ifndef SELF_PROFILE_H
#define SELF_PROFILE_H

#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <time.h>

#ifdef _SELF_PROFILE_VFTRACE
extern char *vftr_self_prof_filename;
extern FILE *vftr_self_prof_iohandle;
extern char *vftr_self_prof_iobuffer;

#define GET_SELF_PROF_FILENAME \
   do { \
      int pid = getpid(); \
      char *filename_base = "vftrace_self_profile-"; \
      int total_str_len = snprintf(NULL, 0, "%s%d", filename_base, pid) + 1; \
      vftr_self_prof_filename = (char*) malloc(total_str_len*sizeof(char)); \
      snprintf(vftr_self_prof_filename, \
               total_str_len, "%s%d", \
               filename_base, pid); \
   } while(0)

#define GET_SELF_PROF_FILEHANDLE \
   do { \
      vftr_self_prof_iohandle = \
         fopen(vftr_self_prof_filename, "w"); \
      size_t bufsize = 256*1024*1024; \
      vftr_self_prof_iobuffer = (char*) malloc(bufsize); \
      memset((void*) vftr_self_prof_iobuffer, 0, bufsize); \
      int status = setvbuf(vftr_self_prof_iohandle, \
                           vftr_self_prof_iobuffer, \
                           _IOFBF, bufsize); \
   } while(0)

#define INIT_SELF_PROF_VFTRACE \
   do { \
      GET_SELF_PROF_FILENAME; \
      GET_SELF_PROF_FILEHANDLE; \
   } while(0)

#define FREE_SELF_PROF_FILENAME \
   do { \
      if (vftr_self_prof_filename != NULL) { \
         free(vftr_self_prof_filename); \
         vftr_self_prof_filename = NULL; \
      } \
   } while(0)

#define CLOSE_SELF_PROF_FILEHANDLE \
   do { \
      if (vftr_self_prof_iohandle != NULL) { \
         fclose(vftr_self_prof_iohandle); \
         vftr_self_prof_iohandle = NULL; \
         free(vftr_self_prof_iobuffer); \
         vftr_self_prof_iobuffer = NULL; \
      } \
   } while(0)
   
#define FINALIZE_SELF_PROF_VFTRACE \
   do { \
      CLOSE_SELF_PROF_FILEHANDLE; \
      FREE_SELF_PROF_FILENAME; \
   } while(0)

#define GET_TIMESTAMP \
   struct timespec timestamp; \
   clock_gettime(CLOCK_MONOTONIC, &timestamp);

#define SELF_PROFILE_START_FUNCTION \
   do { \
      GET_TIMESTAMP; \
      fprintf(vftr_self_prof_iohandle, \
              "Enter: %s at %ld s %ld ns\n", \
              __FUNCTION__, \
              timestamp.tv_sec, \
              timestamp.tv_nsec); \
   } while(0)
   

#define SELF_PROFILE_END_FUNCTION \
   do { \
      GET_TIMESTAMP; \
      fprintf(vftr_self_prof_iohandle, \
              "Leave: %s at %ld s %ld ns\n", \
              __FUNCTION__, \
              timestamp.tv_sec, \
              timestamp.tv_nsec); \
   } while(0)

#else

#define INIT_SELF_PROF_VFTRACE
#define SELF_PROFILE_START_FUNCTION
#define SELF_PROFILE_END_FUNCTION
#define FINALIZE_SELF_PROF_VFTRACE

#endif // ifdef SELF_PROFILE_VFTRACE

#endif // define SELF_PROFILE_H
