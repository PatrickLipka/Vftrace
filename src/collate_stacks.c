#include <stdlib.h>

#include <string.h>
#ifdef _MPI
#include <mpi.h>
#endif

#include "stack_types.h"
#include "collated_hash_types.h"
#include "collated_stack_types.h"

#include "hashing.h"
#include "collate_hashes.h"
#include "search.h"

collated_stacktree_t vftr_new_collated_stacktree() {
   collated_stacktree_t stacktree;
   stacktree.nstacks = 0;
   stacktree.stacks = NULL;
   return stacktree;
}

#ifdef _MPI
void vftr_broadcast_collated_stacktree_root(collated_stacktree_t *stacktree_ptr) {
   int nranks;
   PMPI_Comm_size(MPI_COMM_WORLD, &nranks);
   int nstacks = stacktree_ptr->nstacks;
   // broadcasting the caller ids
   int *tmpintarr = (int*) malloc((nstacks+1)*sizeof(int));
   for (int istack=0; istack<nstacks; istack++) {
      tmpintarr[istack] = stacktree_ptr->stacks[istack].caller;
   }
   PMPI_Bcast(tmpintarr, nstacks, MPI_INT, 0, MPI_COMM_WORLD);
   // communicate the precise-tracing status for each function
   for (int istack=0; istack<nstacks; istack++) {
      tmpintarr[istack] = stacktree_ptr->stacks[istack].precise ? 1 : 0;
   }
   PMPI_Bcast(tmpintarr, nstacks, MPI_INT, 0, MPI_COMM_WORLD);
   // broadcasting the function names.
   // first the length of each name
   int totallen = 0;
   for (int istack=0; istack<nstacks; istack++) {
      tmpintarr[istack] = strlen(stacktree_ptr->stacks[istack].name);
      totallen += tmpintarr[istack];
      totallen += 1; // null terminator between all strings
   }
   // the total length is appended to the tmpintarr
   tmpintarr[nstacks] = totallen;
   PMPI_Bcast(tmpintarr, nstacks+1, MPI_INT, 0, MPI_COMM_WORLD);
   char *tmpchararr = (char*) malloc(totallen*sizeof(char));
   char *charptr = tmpchararr;
   for (int istack=0; istack<nstacks; istack++) {
      strcpy(charptr, stacktree_ptr->stacks[istack].name);
      charptr += tmpintarr[istack];
      charptr++; // null terminator should be copied by strcpy
   }
   PMPI_Bcast(tmpchararr, totallen, MPI_CHAR, 0, MPI_COMM_WORLD);
   free(tmpintarr);
   free(tmpchararr);
}

void vftr_broadcast_collated_stacktree_receivers(collated_stacktree_t *stacktree_ptr) {
   int nranks;
   PMPI_Comm_size(MPI_COMM_WORLD, &nranks);
   int nstacks = stacktree_ptr->nstacks;
   // receiving the caller ids
   int *tmpintarr = (int*) malloc((nstacks+1)*sizeof(int));
   PMPI_Bcast(tmpintarr, nstacks, MPI_INT, 0, MPI_COMM_WORLD);
   for (int istack=0; istack<nstacks; istack++) {
      stacktree_ptr->stacks[istack].caller = tmpintarr[istack];
   }
   // recive the precise-tracing status for each function
   PMPI_Bcast(tmpintarr, nstacks, MPI_INT, 0, MPI_COMM_WORLD);
   for (int istack=0; istack<nstacks; istack++) {
      stacktree_ptr->stacks[istack].precise = tmpintarr[istack] == 1 ? true : false;
   }
   // receiving the function names;
   // first the length of each name;
   int totallen = 0;
   PMPI_Bcast(tmpintarr, nstacks+1, MPI_INT, 0, MPI_COMM_WORLD);
   // the total length is appended to the tmpintarr
   totallen = tmpintarr[nstacks];
   char *tmpchararr = (char*) malloc(totallen*sizeof(char));
   char *charptr = tmpchararr;
   PMPI_Bcast(tmpchararr, totallen, MPI_CHAR, 0, MPI_COMM_WORLD);
   for (int istack=0; istack<nstacks; istack++) {
      stacktree_ptr->stacks[istack].name = strdup(charptr);
      charptr += tmpintarr[istack];
      charptr++; // null terminator should be copied by strcpy
   }
   free(tmpintarr);
   free(tmpchararr);
}

void vftr_broadcast_collated_stacktree(collated_stacktree_t *stacktree_ptr) {
   int nranks;
   PMPI_Comm_size(MPI_COMM_WORLD, &nranks);
   if (nranks > 1) {
      int myrank;
      PMPI_Comm_rank(MPI_COMM_WORLD, &myrank);
      if (myrank == 0) {
         vftr_broadcast_collated_stacktree_root(stacktree_ptr);
      } else {
         vftr_broadcast_collated_stacktree_receivers(stacktree_ptr);
      }
   }
}
#endif

collated_stacktree_t vftr_collate_stacks(stacktree_t *stacktree_ptr) {
   collated_stacktree_t coll_stacktree;

   // first compute the hashes for all stacks
   vftr_compute_stack_hashes(stacktree_ptr);

   // collate hashes between processes
   hashlist_t hashlist = vftr_collate_hashes(stacktree_ptr);

   // create empty collated stacktree
   coll_stacktree.nstacks = hashlist.nhashes;
   coll_stacktree.stacks = (collated_stack_t*)
      malloc(coll_stacktree.nstacks*sizeof(collated_stack_t));
   for (int istack=0; istack<coll_stacktree.nstacks; istack++) {
      coll_stacktree.stacks[istack].local_stack = NULL;
      coll_stacktree.stacks[istack].gid = istack;
      coll_stacktree.stacks[istack].precise = false;
      coll_stacktree.stacks[istack].caller = -1;
      coll_stacktree.stacks[istack].name = NULL;
      coll_stacktree.stacks[istack].hash = hashlist.hashes[istack];
   }

   // build a lookup table each to translate local2global and global2local
   int *local2global_ID = (int*) malloc(stacktree_ptr->nstacks*sizeof(int));
   int *global2local_ID = (int*) malloc(coll_stacktree.nstacks*sizeof(int));
   for (int istack=0; istack<stacktree_ptr->nstacks; istack++) {
      // -1 in the lookup table means that the local stack does not exist
      local2global_ID[istack] = -1;
   }
   for (int istack=0; istack<coll_stacktree.nstacks; istack++) {
      // -1 in the lookup table means that the local stack does not exist
      global2local_ID[istack] = -1;
   }
   // assign every function its global ID
   // the global ID is the index in the hash table
   // fill the looup tables
   for (int istack=0; istack<stacktree_ptr->nstacks; istack++) {
      int idx = vftr_binary_search_uint64(hashlist.nhashes,
                                          hashlist.hashes,
                                          stacktree_ptr->stacks[istack].hash);
      if (idx >= 0) {
         stacktree_ptr->stacks[istack].gid = idx;
         local2global_ID[istack] = idx;
         global2local_ID[idx] = istack;
      }
   }

   // hashtable is no longer needed
   vftr_collated_hashlist_free(&hashlist);

   // fill in the locally known information
   // special treatment for init because it does not have a caller
   int myrank = 0;
#ifdef _MPI
   int mpi_initialized;
   PMPI_Initialized(&mpi_initialized);
   if (mpi_initialized) {
      PMPI_Comm_rank(MPI_COMM_WORLD, &myrank);
   }
#endif
   if (myrank == 0) {
      {
         int lid = 0;
         stack_t *local_stack = stacktree_ptr->stacks+lid;
         int gid = local2global_ID[lid];
         collated_stack_t *global_stack = coll_stacktree.stacks+gid;
         global_stack->local_stack = local_stack;
         global_stack->gid = gid;
         global_stack->precise = local_stack->precise;
         global_stack->caller = -1;
         global_stack->name = strdup(local_stack->name);
      }
      for (int lid=1; lid<stacktree_ptr->nstacks; lid++) {
         stack_t *local_stack = stacktree_ptr->stacks+lid;
         int gid = local2global_ID[lid];
         collated_stack_t *global_stack = coll_stacktree.stacks+gid;
         global_stack->local_stack = local_stack;
         global_stack->gid = gid;
         global_stack->precise = local_stack->precise;
         global_stack->caller = local2global_ID[local_stack->caller];
         global_stack->name = strdup(local_stack->name);
      }
   }

#ifdef _MPI
   if (mpi_initialized) {
      if (myrank == 0) {
         // if there are multiple processes the table might still be missing entries
         int *missingstacks = (int*) malloc(coll_stacktree.nstacks*sizeof(int));
         // loop over all ranks and collect the missing information
         int nranks;
         PMPI_Comm_size(MPI_COMM_WORLD, &nranks);
         for (int irank=1; irank<nranks; irank++) {
            int nmissing = 0;
            for (int istack=0; istack<coll_stacktree.nstacks; istack++) {
               if (coll_stacktree.stacks[istack].name == NULL) {
                  missingstacks[nmissing] = istack;
                  nmissing++;
               }
            }

            // Send to the selected process how many entries are still missing
            PMPI_Send(&nmissing, 1, MPI_INT, irank, 0, MPI_COMM_WORLD);
            // if at least one entry is missing proceed
            if (nmissing > 0) {
               // Send the missing IDs
               PMPI_Send(missingstacks, nmissing, MPI_INT, irank, 0, MPI_COMM_WORLD);
               // Receive how many missing stacks the other process can fill in
               MPI_Status mystat;
               int hasnmissing;
               PMPI_Recv(&hasnmissing, 1, MPI_INT, irank, 0, MPI_COMM_WORLD, &mystat);
               // only proceed if the number of stacks is positive
               if (hasnmissing > 0) {
                   // Allocate space for the Stack information
                   // globalIDs  -> 4*istack+0
                   // returnID   -> 4*istack+1
                   // NameLength -> 4*istack+2
                   // precise    -> 4*istack+3
                   int *missingStackInfo = (int*) malloc(4*hasnmissing*sizeof(int));
                   // Receive the found information from remote process
                   PMPI_Recv(missingStackInfo, 4*hasnmissing, MPI_INT,
                             irank, 0, MPI_COMM_WORLD, &mystat);

                   // Create a buffer that contains all stack names in contatenated form
                   int sumlength = 0;
                   for (int istack=0; istack<hasnmissing; istack++) {
                      sumlength += missingStackInfo[4*istack+2];
                   }
                   char *concatNames = (char*) malloc(sumlength*sizeof(char));

                   // Receive the concatenated String
                   PMPI_Recv(concatNames, sumlength, MPI_CHAR,
                             irank, 0, MPI_COMM_WORLD, &mystat);

                   // Write all the gathered info to the global stackinfo
                   char *tmpstrptr = concatNames;
                   for (int istack=0; istack<hasnmissing; istack++) {
                      int globID = missingStackInfo[4*istack+0];
                      coll_stacktree.stacks[globID].caller = missingStackInfo[4*istack+1];
                      coll_stacktree.stacks[globID].name = strdup(tmpstrptr);
                      // next string
                      tmpstrptr += missingStackInfo[4*istack+2];
                      // is precise ?
                      coll_stacktree.stacks[globID].precise = missingStackInfo[4*istack+3] != 0;
                   }

                   free(concatNames);
                   free(missingStackInfo);

               }
            }
         }
         free(missingstacks);
      } else {
         // not process 0
         MPI_Status mystat;
         // receive how many entries process 0 is missing
         int nmissing;
         PMPI_Recv(&nmissing, 1, MPI_INT, 0, 0, MPI_COMM_WORLD, &mystat);
         // if at least one entry is missing proceed
         if (nmissing > 0) {
            // allocate space to hold the missing ids
            int *missingstacks = (int*) malloc(nmissing*sizeof(int));
            // receiving the missing stacks
            PMPI_Recv(missingstacks, nmissing, MPI_INT, 0, 0, MPI_COMM_WORLD, &mystat);

            // check how many of the missing stacks this process has infos about
            int hasnmissing = 0;
            for (int imissing=0; imissing<nmissing; imissing++) {
               int globID = missingstacks[imissing];
               if (global2local_ID[globID] >= 0) {
                  hasnmissing++;
               }
            }
            // Report back how many missing stacks this process can fill in
            PMPI_Send(&hasnmissing, 1, MPI_INT, 0, 0, MPI_COMM_WORLD);
            // only proceed if the number of stacks is positive
            if (hasnmissing > 0) {
               // Allocate space for the global IDs, return ID, and the name lengths
               // globalIDs  -> 4*istack+0
               // returnID   -> 4*istack+1
               // NameLength -> 4*istack+2
               // Precise    -> 4*istack+3
               int *missingStackInfo = (int*) malloc(4*hasnmissing*sizeof(int));
               // Go through the stacks and record the needed information
               int imatch = 0;
               for (int istack=0; istack<nmissing; istack++) {
                  int globID = missingstacks[istack];
                  int locID = global2local_ID[globID];
                  if (locID >= 0) {
                     missingStackInfo[4*imatch+0] = globID;
                     stack_t *local_stack = stacktree_ptr->stacks+locID;
                     missingStackInfo[4*imatch+1] = local2global_ID[local_stack->caller];
                     // add one to length due to null terminator
                     missingStackInfo[4*imatch+2] = strlen(local_stack->name) + 1;
                     missingStackInfo[4*imatch+3] = local_stack->precise ? 1 : 0;
                     imatch++;
                  }
               }
               // Communicate the found information to process 0;
               PMPI_Send(missingStackInfo, 4*hasnmissing, MPI_INT, 0, 0, MPI_COMM_WORLD);
               // Create a buffer that contains all stack names in contatenated form
               int sumlength = 0;
               for (int istack=0; istack<hasnmissing; istack++) {
                  sumlength += missingStackInfo[4*istack+2];
               }
               char *concatNames = (char*) malloc(sumlength*sizeof(char));
               // concatenate the names into one string
               char *tmpstrptr = concatNames;
               for (int istack=0; istack<hasnmissing; istack++) {
                  int globID = missingStackInfo[4*istack+0];
                  int locID = global2local_ID[globID];
                  strcpy(tmpstrptr, stacktree_ptr->stacks[locID].name);
                  tmpstrptr += missingStackInfo[4*istack+2]-1;
                  // add null terminator
                  *tmpstrptr = '\0';
                  tmpstrptr++;
               }
               // communicate the concatenated string to process 0
               PMPI_Send(concatNames, sumlength, MPI_CHAR, 0, 0, MPI_COMM_WORLD);

               // free everything. This should be all on the remote processes
               free(concatNames);
               free(missingStackInfo);
            }
            free(missingstacks);
         }
      }

      vftr_broadcast_collated_stacktree(&coll_stacktree);
   }
#endif

   free(local2global_ID);
   free(global2local_ID);

   return coll_stacktree;
}

void vftr_collated_stacktree_free(collated_stacktree_t *stacktree_ptr) {
   if (stacktree_ptr->nstacks > 0) {
      for (int istack=0; istack<stacktree_ptr->nstacks; istack++) {
         free(stacktree_ptr->stacks[istack].name);
      }
      free(stacktree_ptr->stacks);
      stacktree_ptr->stacks = NULL;
      stacktree_ptr->nstacks = 0;
   }
}

char *vftr_get_collated_stack_string(collated_stacktree_t stacktree, int stackid) {
   int stringlen = 0;
   int tmpstackid = stackid;
   stringlen += strlen(stacktree.stacks[stackid].name);
   stringlen ++; // function seperating character "<", or null terminator
//   if (stacktree.stacks[tmpstackid].precise) {
//      stringlen ++; // '*' for indicating precise functions
//   }
   while (stacktree.stacks[tmpstackid].caller >= 0) {
      tmpstackid = stacktree.stacks[tmpstackid].caller;
      stringlen += strlen(stacktree.stacks[tmpstackid].name);
      stringlen ++; // function seperating character "<", or null terminator
//      if (stacktree.stacks[tmpstackid].precise) {
//         stringlen ++; // '*' for indicating precise functions
//      }
   }
   char *stackstring = (char*) malloc(stringlen*sizeof(char));
   // copy the chars one by one so there is no need to call strlen again.
   // thus minimizing reading the same memory locations over and over again.
   tmpstackid = stackid;
   char *tmpname_ptr = stacktree.stacks[tmpstackid].name;
   char *tmpstackstring_ptr = stackstring;
   while (*tmpname_ptr != '\0') {
      *tmpstackstring_ptr = *tmpname_ptr;
      tmpstackstring_ptr++;
      tmpname_ptr++;
   }
//   if (stacktree.stacks[tmpstackid].precise) {
//      *tmpstackstring_ptr = '*';
//      tmpstackstring_ptr++;
//   }
   while (stacktree.stacks[tmpstackid].caller >= 0) {
      // add function name separating character
      *tmpstackstring_ptr = '<';
      tmpstackstring_ptr++;
      tmpstackid = stacktree.stacks[tmpstackid].caller;
      char *tmpname_ptr = stacktree.stacks[tmpstackid].name;
      while (*tmpname_ptr != '\0') {
         *tmpstackstring_ptr = *tmpname_ptr;
         tmpstackstring_ptr++;
         tmpname_ptr++;
      }
//      if (stacktree.stacks[tmpstackid].precise) {
//         *tmpstackstring_ptr = '*';
//         tmpstackstring_ptr++;
//      }
   }
   // replace last char with a null terminator
   *tmpstackstring_ptr = '\0';
   return stackstring;
}

void vftr_print_collated_stack(FILE *fp, collated_stacktree_t stacktree, int stackid) {
   char *stackstr = vftr_get_collated_stack_string(stacktree, stackid);
   fprintf(fp, "%s", stackstr);
   free(stackstr);
}

void vftr_print_collated_stacklist(FILE *fp, collated_stacktree_t stacktree) {
   for (int istack=0; istack<stacktree.nstacks; istack++) {
      fprintf(fp, "%u: ", istack);
      vftr_print_collated_stack(fp, stacktree, istack);
      fprintf(fp, "\n");
   }
}
