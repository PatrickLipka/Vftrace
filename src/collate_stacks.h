#ifndef COLLATE_STACKS_H
#define COLLATE_STACKS_H

#include "stack_types.h"
#include "collated_stack_types.h"

collated_stacktree_t vftr_new_collated_stacktree();

collated_stacktree_t vftr_collate_stacks(stacktree_t *stacktree_ptr);

void vftr_collated_stacktree_free(collated_stacktree_t *stacktree_ptr);

void vftr_print_collated_stacklist(FILE *fp, collated_stacktree_t stacktree);

char *vftr_get_collated_stack_string(collated_stacktree_t stacktree, int stackid);

#endif
