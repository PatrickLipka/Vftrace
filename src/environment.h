#ifndef ENVIRONMENT_H
#define ENVIRONMENT_H

#include <stdio.h>

#include "environment_types.h"

env_var_t *vftr_get_env_var_ptr_by_idx(environment_t *environment_ptr, int idx);

void vftr_print_environment(FILE *fp, environment_t environment);

environment_t vftr_read_environment();

void vftr_check_env_names(FILE *fp, environment_t *environment_ptr);

void vftr_environment_free(environment_t *environment_ptr);

void vftr_environment_assert(FILE *fp, environment_t environment);

#endif
