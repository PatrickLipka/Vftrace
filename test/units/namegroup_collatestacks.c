#include <stdlib.h>
#include <stdio.h>

#include "self_profile.h"
#include "environment_types.h"
#include "environment.h"
#include "symbol_types.h"
#include "symbols.h"
#include "stack_types.h"
#include "stacks.h"
#include "collated_stack_types.h"
#include "collate_stacks.h"

#include "dummysymboltable.h"

#ifdef _MPI
#include <mpi.h>
#endif

int main(int argc, char **argv) {
   INIT_SELF_PROF_VFTRACE;
#if defined(_MPI)
   PMPI_Init(&argc, &argv);
#else
   (void) argc;
   (void) argv;
#endif

   // dummy symboltable
   uintptr_t addrs = 123456;
   symboltable_t symboltable = dummy_symbol_table(6, addrs);

   // build stacktree
   stacktree_t stacktree = vftr_new_stacktree();

   char *name;
   int idx0 = 0;
   // func0<init
   name = vftr_get_name_from_address(symboltable, addrs+0);
   int idx1 = vftr_new_stack(idx0, &stacktree,
                             name, name, addrs+0, false);
   // func2<func0<init
   name = vftr_get_name_from_address(symboltable, addrs+2);
   int idx2 = vftr_new_stack(idx1, &stacktree,
                             name, name, addrs+2, false);

   // func1<init
   name = vftr_get_name_from_address(symboltable, addrs+1);
   int idx3 = vftr_new_stack(idx0, &stacktree,
                             name, name, addrs+1, false);
   // func2<func1<init
   name = vftr_get_name_from_address(symboltable, addrs+2);
   int idx4 = vftr_new_stack(idx3, &stacktree,
                             name, name, addrs+2, false);
   // func3<func1<init
   name = vftr_get_name_from_address(symboltable, addrs+3);
   int idx5 = vftr_new_stack(idx3, &stacktree,
                             name, name, addrs+3, false);
   // func4<func1<init
   name = vftr_get_name_from_address(symboltable, addrs+4);
   int idx6 = vftr_new_stack(idx3, &stacktree,
                             name, name, addrs+4, false);

   // func5<init
   name = vftr_get_name_from_address(symboltable, addrs+5);
   int idx7 = vftr_new_stack(idx0, &stacktree,
                             name, name, addrs+5, false);
   // func3<func5<init
   name = vftr_get_name_from_address(symboltable, addrs+3);
   int idx8 = vftr_new_stack(idx7, &stacktree,
                             name, name, addrs+3, false);
   // func4<func5<init
   name = vftr_get_name_from_address(symboltable, addrs+4);
   int idx9 = vftr_new_stack(idx7, &stacktree,
                             name, name, addrs+4, false);
   // func2<func4<func5<init
   name = vftr_get_name_from_address(symboltable, addrs+2);
   int idx10 = vftr_new_stack(idx9, &stacktree,
                              name, name, addrs+2, false);
   // func3<func4<func5<init
   name = vftr_get_name_from_address(symboltable, addrs+3);
   int idx11 = vftr_new_stack(idx9, &stacktree,
                              name, name, addrs+3, false);

   collated_stacktree_t collated_stacktree = vftr_collate_stacks(&stacktree);
   vftr_print_collated_stacklist(stdout, collated_stacktree);

   fprintf(stdout, "\n");
   collated_stacktree_t namegrouped_collated_stacktree =
      vftr_collated_stacktree_group_by_name(&collated_stacktree);
   vftr_print_collated_stacklist(stdout, namegrouped_collated_stacktree);

   free_dummy_symbol_table(&symboltable);
   vftr_stacktree_free(&stacktree);
   vftr_collated_stacktree_free(&collated_stacktree);
   vftr_collated_stacktree_free(&namegrouped_collated_stacktree);

#ifdef _MPI
   PMPI_Finalize();
#endif

   FINALIZE_SELF_PROF_VFTRACE;
   return 0;
}
