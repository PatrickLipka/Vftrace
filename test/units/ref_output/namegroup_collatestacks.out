0: func2<func1<init
1: func4<func1<init
2: func3<func4<func5<init
3: func1<init
4: func3<func1<init
5: func2<func0<init
6: func5<init
7: init
8: func4<func5<init
9: func0<init
10: func3<func5<init
11: func2<func4<func5<init

0: func0: 9
1: func1: 3
2: func2: 0,5,11
3: func3: 2,4,10
4: func4: 1,8
5: func5: 6
6: init: 7
